### PR TITLE
SNS - GPIO (read/write)

### PR DESCRIPTION
**Pull request by SnickeyX**
This PR provides an (alternative) implementation of GPIO using read/write only. Using mmap provides a faster(* in certain scenarios) implementation but I want to test if the difference is significant at all between the two. 

Goal:
To test the real difference in I/O between mmap v read/write during a test run (with our car).

### CODE REVIEW

File: lib/io/hardware_gpio.cpp [hunk 1]
```diff
@@ -2,13 +2,13 @@
...
 
-std::optional<core::DigitalSignal> HardwareGpioReader::read()
+std::optional<core::DigitalSignal> HardwareGpioReader::readPin()
 {
   // TODOLater: implement
   throw -1;
 }
 
-core::Result HardwareGpioWriter::write(const core::DigitalSignal state)
+core::Result HardwareGpioWriter::writeToPin(const core::DigitalSignal state)
```
Comment (mifrandir) [inline lib/io/hardware_gpio.cpp:11]: not sure i agree with this rename? this function doesn't take a pin as an argument so why is it called `writeToPin`?

File: lib/io/hardware_gpio_alt.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    logger_.log(core::LogLevel::kFatal, "Failed to initialise GPIO %d", pin);
+    return std::nullopt;
+  }
+  const int read_file_descriptor = getFileDescriptor(pin, Direction::kIn);
+  if (read_file_descriptor < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to get file descriptor for GPIO %d", pin);
+    return std::nullopt;
+  }
+  HardwareGpioReader reader(logger_, read_file_descriptor);
+  return std::make_shared<HardwareGpioReader>(reader);
```
Comment (mifrandir) [inline lib/io/hardware_gpio_alt.cpp:90]: i want to comment on something here, for your understanding; feel free to resolve after reading (RAR)

The "right" way to do this would be `return std::make_shared<...>(logger_, read_file_descriptor);`. Of course, this doesn't work because the constructor is private so that wouldn't work. 

How could we resolve this? Option 1 is to use nested classes and make the reader/write a private class inside the Gpio. Then we could make the constructor public (and within the `HardwareGpio` class). We like to avoid nested classes so we accept the ugliness of this. Option 2 would be to make the constructor public. This usually isn't a problem if we assume that the user understands the design. However, this is a bold assumption.

Besides style, why would the single line way be better? Well, we avoid an unnecessary copy. Then again, these sorts of functions are only called on initialisation (i.e. O(1) times) so we don't really care about performance here.

**Review by mifrandir — CHANGES_REQUESTED** [review]
didn't read in too much detail as it's experimental anyways

File: lib/io/hardware_gpio.cpp (hunk 1 above)
Comment (SnickeyX) [reply lib/io/hardware_gpio.cpp:11]: `writeState` and `readState`?

Comment (SnickeyX) [reply lib/io/hardware_gpio.cpp:11]: or signal instead of state

File: lib/io/hardware_gpio_alt.cpp (hunk 2 above)
Comment (SnickeyX) [reply lib/io/hardware_gpio_alt.cpp:90]: i see, do appreciate this btw. not always sure what the "right" way to do things even are

File: lib/io/hardware_gpio.cpp (hunk 1 above)
Comment (mifrandir) [reply lib/io/hardware_gpio.cpp:11]: but like... why not just `write`? I get that we want to make things descriptive but `brakes_pin_.write(core::DigitalSignal::kHigh)` doesn't look particularly hard to parse, for example?

Comment (SnickeyX) [reply lib/io/hardware_gpio.cpp]: because the compiler confuses the linux `write` with the gpio `write` and thinks its a recursive function. not sure how to get around it with c libraries not having a namepsace -> was what I was thinking, just found out that global namespace is a thing

**Review by mifrandir — APPROVED** [review]

**Review by kshxtij — APPROVED** [review]
